### PR TITLE
#2276 implement SmbFileSystem and IT test

### PR DESCRIPTION
**Pull request by eechih**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #2276 )
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-client/src/main/scala/com/island/ohara/client/filesystem/smb/SmbClient.scala [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+import com.hierynomus.smbj.share.DiskShare
+import com.hierynomus.smbj.{SMBClient, SmbConfig}
+import com.island.ohara.client.filesystem.{FileSystem, FileType}
+import com.island.ohara.common.annotations.Optional
+import com.island.ohara.common.exception.OharaSmbException
+import com.island.ohara.common.util.{CommonUtils, Releasable}
+
+import scala.collection.JavaConverters._
+
+trait SmbClient extends FileSystem
```
Comment (chia7712) [inline ohara-client/src/main/scala/com/island/ohara/client/filesystem/smb/SmbClient.scala]: Could we rename SmbClient to SmbFileSystem?

Comment (chia7712) [reply ohara-client/src/main/scala/com/island/ohara/client/filesystem/smb/SmbClient.scala]: Or we offer builder for smb than have a explicit class called SmbClient

**Comment by chia7712** [comment]
what the difference between FileSystem and Storage?

**Comment by eechih** [comment]
> what the difference between FileSystem and Storage?

The purpose of creating `FileSystem` trait is to standardize the client implementation of those file system, like `FtpClient`, `SmbClient` etc. So I will fire an issue to refactor `FtpClient`. 

It seems that `FileSystem` and `Storage` are very similar, so it is possible to keep only one of them in the future.

File: ohara-client/src/main/scala/com/island/ohara/client/filesystem/smb/SmbClient.scala (hunk 1 above)
Comment (eechih) [reply ohara-client/src/main/scala/com/island/ohara/client/filesystem/smb/SmbClient.scala]: > Or we offer builder for smb than have a explicit class called SmbClient

We already have a builder method for creating `SmbClient`.

```
SmbClient
        .builder()
        .hostname(hostname.get)
        .port(Integer.parseInt(port.get))
        .user(user.get)
        .password(password.get)
        .shareName(shareName.get)
        .build)
```

Comment (chia7712) [reply ohara-client/src/main/scala/com/island/ohara/client/filesystem/smb/SmbClient.scala]: I meant we don't have a class called SmbClient.

For example, FileSystem.smbBuilder

**Comment by chia7712** [comment]
> It seems that FileSystem and Storage are very similar, so it is possible to keep only one of them in the future.

It would be better to integrate them into a single class if they are similar ...

File: ohara-client/src/main/scala/com/island/ohara/client/filesystem/smb/SmbClient.scala (hunk 1 above)
Comment (eechih) [reply ohara-client/src/main/scala/com/island/ohara/client/filesystem/smb/SmbClient.scala]: Oh~ I misunderstood, I will fix it. Thanks. :)

Comment (eechih) [reply ohara-client/src/main/scala/com/island/ohara/client/filesystem/smb/SmbClient.scala]: @chia7712 address your comment, please review again, thanks.

**Comment by eechih** [comment]
> It would be better to integrate them into a single class if they are similar ...

This task will be done in another issue #2572 

@chia7712 Thanks you for your assistance.

**Comment by jackyoh** [comment]
retry connector

**Comment by jackyoh** [comment]
retry it

**Comment by eechih** [comment]
retry otherIT

**Comment by eechih** [comment]
retry otherIT

**Pull request merged**